### PR TITLE
W5 - PhongLe - Thread - Handle - CountdownTimer

### PR DESCRIPTION
**Pull request by at-phongle**
#### Related issues (Optional)
Task #66

#### Description (Required)
- [x] Thread Handler
- [x] CountdownTimer

### CODE REVIEW

File: app/src/at-phongle/java/vn/asiantech/internship/asynchronous/CountdownTimerActivity.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        setContentView(R.layout.activity_countdown);
+        mRecyclerViewCountdown = findViewById(R.id.recyclerViewCountdown);
+        initData();
+        initAdapter();
+        new CountDownTimer(180000, 5000) {
+            @Override
+            public void onTick(long millisUntilFinished) {
+                ++mCount;
+                if (mCount % 2 == 0) {
+                    mListItem.add("Item " + String.valueOf(mListItem.size() + 1));
```
Comment (at-dev-support-01) [inline app/src/at-phongle/java/vn/asiantech/internship/asynchronous/CountdownTimerActivity.java]: WARNING: [UselessStringValueOf] No need to call String.valueOf to append to a string.
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/strings.html#UselessStringValueOf

File: app/src/at-phongle/java/vn/asiantech/internship/asynchronous/CountdownTimerActivity.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+                    mListItem.add("Item " + String.valueOf(mListItem.size() + 1));
```
Comment (at-dev-support-01) [inline app/src/at-phongle/java/vn/asiantech/internship/asynchronous/CountdownTimerActivity.java]: WARNING: [UselessStringValueOf] No need to call String.valueOf to append to a string.
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/strings.html#UselessStringValueOf

File: app/src/at-phongle/java/vn/asiantech/internship/asynchronous/DownloadImageFragment.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+                    @Override
+                    public void run() {
+                        float percent = (float) mTotal / (float) mPhotoLength * 100;
+                        updateProgress((int) percent);
+                    }
+                });
+                output.write(data, 0, count);
+            }
+        } catch (Throwable e) {
+            e.printStackTrace();
```
Comment (at-dev-support-01) [inline app/src/at-phongle/java/vn/asiantech/internship/asynchronous/DownloadImageFragment.java]: WARNING: [AvoidPrintStackTrace] Avoid printStackTrace(); use a logger call instead.
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/logging-java.html#AvoidPrintStackTrace

File: app/src/at-phongle/java/vn/asiantech/internship/asynchronous/DownloadImageFragment.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        } finally {
+            try {
+                if (connection != null)
```
Comment (at-dev-support-01) [inline app/src/at-phongle/java/vn/asiantech/internship/asynchronous/DownloadImageFragment.java]: WARNING: [IfStmtsMustUseBraces] Avoid using if statements without curly braces
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/braces.html#IfStmtsMustUseBraces

File: app/src/at-phongle/java/vn/asiantech/internship/asynchronous/DownloadImageFragment.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+                    connection.disconnect();
+                if (is != null)
```
Comment (at-dev-support-01) [inline app/src/at-phongle/java/vn/asiantech/internship/asynchronous/DownloadImageFragment.java]: WARNING: [IfStmtsMustUseBraces] Avoid using if statements without curly braces
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/braces.html#IfStmtsMustUseBraces

File: app/src/at-phongle/java/vn/asiantech/internship/asynchronous/DownloadImageFragment.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+                    is.close();
+            } catch (Exception e) {
+                e.printStackTrace();
```
Comment (at-dev-support-01) [inline app/src/at-phongle/java/vn/asiantech/internship/asynchronous/DownloadImageFragment.java]: WARNING: [AvoidPrintStackTrace] Avoid printStackTrace(); use a logger call instead.
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/logging-java.html#AvoidPrintStackTrace

File: app/src/at-phongle/java/vn/asiantech/internship/asynchronous/DownloadImageFragment.java [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+            }
+        }
+        ((HandlerThreadActivity) getActivity()).setmBitmap(mBitmap);
+    }
+
+    private void updateProgress(int percent) {
+        mProgressBarDownload.setProgress(percent);
+        String text = String.valueOf(percent) + getResources().getString(R.string.tv_percen);
```
Comment (at-dev-support-01) [inline app/src/at-phongle/java/vn/asiantech/internship/asynchronous/DownloadImageFragment.java]: WARNING: [UselessStringValueOf] No need to call String.valueOf to append to a string.
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/strings.html#UselessStringValueOf

File: app/src/at-phongle/java/vn/asiantech/internship/asynchronous/DownloadImageFragment.java [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+                    public void run() {
+                        float percent = (float) mTotal / (float) mPhotoLength * 100;
+                        updateProgress((int) percent);
+                    }
+                });
+                output.write(data, 0, count);
+            }
+        } catch (Throwable e) {
+            e.printStackTrace();
+        } finally {
```
Comment (at-dev-support-01) [inline app/src/at-phongle/java/vn/asiantech/internship/asynchronous/DownloadImageFragment.java]: WARNING: [AvoidPrintStackTrace] Avoid printStackTrace(); use a logger call instead.
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/logging-java.html#AvoidPrintStackTrace

**Comment by at-phongle** [comment]
@at-kietvo please review this.thanks

File: app/src/at-phongle/java/vn/asiantech/internship/asynchronous/CountdownTimerAdapter.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+import vn.asiantech.internship.R;
+
+/**
+ * Created by phongle on 21/12/2560.
+ * Adapter for recyclerView Item countdown
+ */
+
```
Comment (at-kietvo) [inline app/src/at-phongle/java/vn/asiantech/internship/asynchronous/CountdownTimerAdapter.java]: remove thiss line

File: app/src/at-phongle/java/vn/asiantech/internship/asynchronous/CountdownTimerAdapter.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    public void onBindViewHolder(ItemViewHolder holder, int position) {
+        holder.onBindData(mItemList, position);
+    }
+
+    @Override
+    public int getItemCount() {
+        return mItemList.size();
+    }
+
+    static class ItemViewHolder extends RecyclerView.ViewHolder {
```
Comment (at-kietvo) [inline app/src/at-phongle/java/vn/asiantech/internship/asynchronous/CountdownTimerAdapter.java]: java doc

File: app/src/at-phongle/java/vn/asiantech/internship/asynchronous/CountdownTimerActivity.java [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+    private int mCount;
+
+    @Override
+    protected void onCreate(@Nullable Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_countdown);
+        mRecyclerViewCountdown = findViewById(R.id.recyclerViewCountdown);
+        initData();
+        initAdapter();
+        new CountDownTimer(180000, 5000) {
```
Comment (at-kietvo) [inline app/src/at-phongle/java/vn/asiantech/internship/asynchronous/CountdownTimerActivity.java]: Pls define the number by `variable`
And then after use countdown, in `onDestroy()` lifecyce of Activity, pls cancel all count down timer

File: app/src/at-phongle/java/vn/asiantech/internship/asynchronous/DownloadImageFragment.java [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+                    connection.disconnect();
+                }
+                if (is != null) {
+                    is.close();
+                }
+            } catch (Exception e) {
+                Log.d("vv", "Download Fail!");
+            }
+        }
+        ((HandlerThreadActivity) getActivity()).setmBitmap(mBitmap);
```
Comment (at-kietvo) [inline app/src/at-phongle/java/vn/asiantech/internship/asynchronous/DownloadImageFragment.java]: Pls check null of Bitmap
Also change mthod name follow convention

File: app/src/at-phongle/java/vn/asiantech/internship/asynchronous/DownloadImageFragment.java [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+            connection = (HttpURLConnection) url.openConnection();
+            is = connection.getInputStream();
+            mBitmap = BitmapFactory.decodeStream(is);
+            // Update progressBar
+            mPhotoLength = connection.getContentLength();
+            InputStream input = new BufferedInputStream(url.openStream(), 8192);
+            OutputStream output = new FileOutputStream(Environment.getExternalStorageDirectory().getPath() + "/downloadedfile.jpg");
+            byte data[] = new byte[1024];
+            int count;
+            while ((count = input.read(data)) != -1) {
```
Comment (at-kietvo) [inline app/src/at-phongle/java/vn/asiantech/internship/asynchronous/DownloadImageFragment.java]: You are limiting the part of the total for reading stream, pls remove data, we can use inputStream.read() instead.
Also remove variable of data[] and change condition of while

File: app/src/at-phongle/java/vn/asiantech/internship/asynchronous/DownloadImageFragment.java [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+        HttpURLConnection connection = null;
+        InputStream is = null;
+        try {
+            URL url = new URL(URL_IMAGE);
+            connection = (HttpURLConnection) url.openConnection();
+            is = connection.getInputStream();
+            mBitmap = BitmapFactory.decodeStream(is);
+            // Update progressBar
+            mPhotoLength = connection.getContentLength();
+            InputStream input = new BufferedInputStream(url.openStream(), 8192);
```
Comment (at-kietvo) [inline app/src/at-phongle/java/vn/asiantech/internship/asynchronous/DownloadImageFragment.java]: Pls close input, output after use

File: app/src/at-phongle/java/vn/asiantech/internship/asynchronous/HandlerThreadActivity.java [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+import android.support.v4.view.ViewPager;
+import android.support.v7.app.AppCompatActivity;
+
+import vn.asiantech.internship.R;
+
+/**
+ * Created by phongle on 19/12/2560.
+ * Activity Handler Thread
+ */
+
```
Comment (at-kietvo) [inline app/src/at-phongle/java/vn/asiantech/internship/asynchronous/HandlerThreadActivity.java]: Pls remove 1 line

File: app/src/at-phongle/java/vn/asiantech/internship/asynchronous/HandlerThreadActivity.java [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+        mViewPager.setAdapter(handlerThreadAdapter);
+        mTabLayout.setupWithViewPager(mViewPager);
+    }
+
+    private void initViews() {
+        mViewPager = findViewById(R.id.viewPagerThreadHandler);
+        mTabLayout = findViewById(R.id.tabLayoutThreadHandler);
+    }
+
+    public Bitmap getmBitmap() {
```
Comment (at-kietvo) [inline app/src/at-phongle/java/vn/asiantech/internship/asynchronous/HandlerThreadActivity.java]: Pls change method name

File: app/src/at-phongle/java/vn/asiantech/internship/asynchronous/HandlerThreadAdapter.java [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+
+import android.support.v4.app.Fragment;
+import android.support.v4.app.FragmentManager;
+import android.support.v4.app.FragmentPagerAdapter;
+
+/**
+ * Created by phongle on 22/12/2560.
+ * Adapter ViewPager Thread Handler
+ */
+
```
Comment (at-kietvo) [inline app/src/at-phongle/java/vn/asiantech/internship/asynchronous/HandlerThreadAdapter.java]: Pls remove thí line

File: app/src/at-phongle/res/layout/activity_app.xml [hunk 18]
```diff
@@ -59,4 +60,14 @@
         android:layout_height="wrap_content"
         android:text="@string/btn_viewpager" />
 
+    <Button
+        android:id="@+id/btnThreadHandler"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:text="@string/btn_thread_handler" />
+    <Button
```
Comment (at-kietvo) [inline app/src/at-phongle/res/layout/activity_app.xml]: Pls reformat code

File: app/src/at-phongle/res/layout/activity_handler_thread.xml [hunk 19]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="utf-8"?>
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:orientation="vertical">
+    <android.support.design.widget.TabLayout
```
Comment (at-kietvo) [inline app/src/at-phongle/res/layout/activity_handler_thread.xml]: Pls reformat code

File: app/src/at-phongle/res/layout/fragment_download_image.xml [hunk 20]
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0" encoding="utf-8"?>
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:orientation="vertical"
+    android:gravity="center">
+    <ProgressBar
```
Comment (at-kietvo) [inline app/src/at-phongle/res/layout/fragment_download_image.xml]: Pls reformat code

**Comment by at-phongle** [comment]
@at-kietvo I checked comment. Please review this. Thanks

File: app/src/at-phongle/java/vn/asiantech/internship/asynchronous/CountdownTimerActivity.java [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+                }
+                if (mCount % 3 == 0) {
+                    mListItem.remove(mListItem.size() / 2);
+                }
+                mCountdownTimerAdapter.notifyDataSetChanged();
+            }
+
+            @Override
+            public void onFinish() {
+                onDestroy();
```
Comment (at-kietvo) [inline app/src/at-phongle/java/vn/asiantech/internship/asynchronous/CountdownTimerActivity.java]: May we miscommunication.
I mean we will cancel CountDownTimer in `onDestroy()` Activity Lifecycle.
It mean do not add method inside onFinish like this.

Pls overrid onDestroy method and cancel count down timer

**Comment by at-phongle** [comment]
@at-kietvo i checked comment. Please review this. Thanks

**Pull request closed without merging**